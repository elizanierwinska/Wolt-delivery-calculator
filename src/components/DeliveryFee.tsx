import '../index.css';
import { useState } from 'react';
import getDeliveryFee from '../calculate';

function DeliveryFee() {
  const [data, setData] = useState({
    cartValue: 0,
    deliveryDistance: 0,
    amountOfItems: 0,
    time: '',
  });
  const [deliveryPrice, setDeliveryPrice] = useState(0);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setData((prevData) => {
      return {
        ...prevData,
        [event.target.name]: event.target.value,
      };
    });
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    setDeliveryPrice(getDeliveryFee(data));
  };

  return (
    <div>
      <form onSubmit={handleSubmit}>
        <div className="flex">
          <p className="flex-title" id="first">
            Cart Value
          </p>
          <input
            type="number"
            onChange={handleChange}
            name="cartValue"
            value={data.cartValue}
            data-testid="cartValue"
          />
          <p className="flex-unit">€</p>
        </div>
        <div className="flex">
          <p className="flex-title">Delivery distance</p>
          <input
            type="number"
            onChange={handleChange}
            name="deliveryDistance"
            value={data.deliveryDistance}
            data-testid="deliveryDistance"
          />
          <p className="flex-unit">m</p>
        </div>
        <div className="flex">
          <p className="flex-title">Amount of items</p>
          <input
            type="number"
            onChange={handleChange}
            name="amountOfItems"
            value={data.amountOfItems}
            data-testid="amountOfItems"
          />
          <p className="flex-unit"></p>
        </div>
        <div>
          <div className="flex">
            <p className="flex-title" id="first">
              Time
            </p>
            <input
              type="date"
              name="time"
              onChange={handleChange}
              value={data.time}
              data-testid="time"
            />
            <p className="flex-unit"></p>
          </div>
        </div>
        <div className="flex-btn">
          <input
            type="submit"
            id="button"
            value="Calculate delivery price"
            data-testid="submitButton"
          />
        </div>
      </form>
      <p id="price" data-testid="fee">
        Delivery price:{' '}
        {deliveryPrice % 1 !== 0 ? deliveryPrice.toFixed(2) : deliveryPrice}€
      </p>
    </div>
  );
}

export default DeliveryFee;
