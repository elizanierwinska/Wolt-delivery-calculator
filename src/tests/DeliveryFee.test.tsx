import { fireEvent, render, screen } from '@testing-library/react';
import * as ShallowRenderer from 'react-test-renderer/shallow';
import DeliveryFee from '../components/DeliveryFee';
import { act } from 'react-test-renderer';
import '@testing-library/jest-dom';

const renderer = ShallowRenderer.createRenderer();

const defaultComponent = <DeliveryFee />;

const setupCartValue = () => {
  const utils = render(defaultComponent);
  const cartInput = screen.getByTestId<HTMLInputElement>('cartValue');
  return {
    cartInput,
    ...utils,
  };
};

const setupDeliveryDistance = () => {
  const utils = render(defaultComponent);
  const deliveryDistanceInput =
    screen.getByTestId<HTMLInputElement>('cartValue');
  return {
    deliveryDistanceInput,
    ...utils,
  };
};

const setupAmountOfItems = () => {
  const utils = render(defaultComponent);
  const amountOfItemsInput = screen.getByTestId<HTMLInputElement>('cartValue');
  return {
    amountOfItemsInput,
    ...utils,
  };
};

const setupTime = () => {
  const utils = render(defaultComponent);
  const timeInput = screen.getByTestId<HTMLInputElement>('time');
  return {
    timeInput,
    ...utils,
  };
};

describe('Testing React Components', () => {
  it('should render and match the snapshot', () => {
    renderer.render(defaultComponent);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });

  test('cartValue.value should change when user is typing', async () => {
    const { cartInput } = setupCartValue();
    fireEvent.change(cartInput, { target: { value: 23 } });

    expect(cartInput.value).toBe('23');
  });

  test('deliveryDistance.value should change when user is typing', async () => {
    const { deliveryDistanceInput } = setupDeliveryDistance();
    fireEvent.change(deliveryDistanceInput, { target: { value: 15230 } });
    expect(deliveryDistanceInput.value).toBe('15230');
  });

  test('amountOfItems.value should change when user is typing', async () => {
    const { amountOfItemsInput } = setupAmountOfItems();
    fireEvent.change(amountOfItemsInput, { target: { value: 17 } });
    expect(amountOfItemsInput.value).toBe('17');
  });

  test('time.value should change when user is typing', async () => {
    const { timeInput } = setupTime();
    fireEvent.click(timeInput, { target: { value: '2024-01-01' } });
    expect(timeInput.value).toBe('2024-01-01');
  });

  test('Valid form data submit', async () => {
    const { getByTestId } = render(defaultComponent);
    const submitButton = getByTestId('submitButton');

    await act(async () => {
      fireEvent.click(submitButton);
    });

    expect(getByTestId('cartValue')).toBeInTheDocument();
    expect(getByTestId('deliveryDistance')).toBeInTheDocument();
    expect(getByTestId('amountOfItems')).toBeInTheDocument();
    expect(getByTestId('time')).toBeInTheDocument();
  });

  test('fee change', async () => {
    const { getByTestId } = render(defaultComponent);
    const submitButton = getByTestId('submitButton');
    const cartValue = getByTestId('cartValue');
    const deliveryDistance = getByTestId('deliveryDistance');
    const amountOfItems = getByTestId('amountOfItems');
    const time = getByTestId('time');

    fireEvent.change(cartValue, { target: { value: 2 } });
    fireEvent.change(deliveryDistance, { target: { value: 1500 } });
    fireEvent.change(amountOfItems, { target: { value: 2 } });
    fireEvent.change(time, { target: { value: '2023-12-14' } });

    await act(async () => {
      fireEvent.click(submitButton);
    });

    expect(getByTestId('fee')).toHaveTextContent('Delivery price: 11â‚¬');
  });
});
