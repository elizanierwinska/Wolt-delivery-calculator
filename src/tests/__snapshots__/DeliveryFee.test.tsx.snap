// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing React Components should render and match the snapshot 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="flex"
    >
      <p
        className="flex-title"
        id="first"
      >
        Cart Value
      </p>
      <input
        data-testid="cartValue"
        name="cartValue"
        onChange={[Function]}
        type="number"
        value={0}
      />
      <p
        className="flex-unit"
      >
        €
      </p>
    </div>
    <div
      className="flex"
    >
      <p
        className="flex-title"
      >
        Delivery distance
      </p>
      <input
        data-testid="deliveryDistance"
        name="deliveryDistance"
        onChange={[Function]}
        type="number"
        value={0}
      />
      <p
        className="flex-unit"
      >
        m
      </p>
    </div>
    <div
      className="flex"
    >
      <p
        className="flex-title"
      >
        Amount of items
      </p>
      <input
        data-testid="amountOfItems"
        name="amountOfItems"
        onChange={[Function]}
        type="number"
        value={0}
      />
      <p
        className="flex-unit"
      />
    </div>
    <div>
      <div
        className="flex"
      >
        <p
          className="flex-title"
          id="first"
        >
          Time
        </p>
        <input
          data-testid="time"
          name="time"
          onChange={[Function]}
          type="date"
          value=""
        />
        <p
          className="flex-unit"
        />
      </div>
    </div>
    <div
      className="flex-btn"
    >
      <input
        data-testid="submitButton"
        id="button"
        type="submit"
        value="Calculate delivery price"
      />
    </div>
  </form>
  <p
    data-testid="fee"
    id="price"
  >
    Delivery price:
     
    0
    €
  </p>
</div>
`;
